MyLibrary/BusinessLogicTests/BookServiceTests.cs

```diff
@@ -0,0 +1,453 @@
+            {
+                new BookAuthorPair { Id = 1, BookId = 1, AuthorId = 2 }
+            };
+            var mockDataProvider = new Mock<IDataProvider>();
+            mockDataProvider.SetupGet(mock => mock.Books).Returns(new List<Book> { book });
+            mockDataProvider.SetupGet(mock => mock.Authors).Returns(new List<Author> { author });
+            mockDataProvider.SetupGet(mock => mock.PairsBookAuthor).Returns(pairBookAuthor);
+            var libraryService = new BookService(mockDataProvider.Object);
+
+            // Act
+            int actualIdOfAddedPair = libraryService.AddAuthorToBook(bookId, authorId);
+
+            // Assert
+            mockDataProvider.Verify(mock => mock.AddBookAuthorPair(It.IsAny<BookAuthorPair>()), Times.Once());
+        }
+
+        /// <summary>
+        /// Tests adding the author to book
+        /// Incorrect input
+        /// Expected exception
+        /// </summary>
+        /// <param name="bookId">The book identifier</param>
+        /// <param name="authorId">The author identifier</param>
+        [TestMethod]
+        [ExpectedException(typeof(ArgumentException))]
+        [DataRow(2, 1)]
+        [DataRow(1, 2)]
+        public void AddAuthorToBook_IncorrectInput_ExpectedException(int bookId, int authorId)
+        {
+            // Arrange
+            Book book = new Book { Id = 1, Name = "Test Book", Year = 1256 };
+            Author author = new Author { Id = 1, FullName = "Test author" };
+            List<BookAuthorPair> pairBookAuthor = new List<BookAuthorPair>()
+            {
+                new BookAuthorPair { Id = 1, BookId = 1, AuthorId = 2 }
+            };
+            var mockDataProvider = new Mock<IDataProvider>();
+            mockDataProvider.SetupGet(mock => mock.Books).Returns(new List<Book> { book });
+            mockDataProvider.SetupGet(mock => mock.Authors).Returns(new List<Author> { author });
+            mockDataProvider.SetupGet(mock => mock.PairsBookAuthor).Returns(pairBookAuthor);
+            var libraryService = new BookService(mockDataProvider.Object);
+
+            // Act
+            int actualIdOfAddedPair = libraryService.AddAuthorToBook(bookId, authorId);
+
+            // Assert
+            mockDataProvider.Verify(mock => mock.AddBookAuthorPair(It.IsAny<BookAuthorPair>()), Times.Never());
+        }
+
+        /// <summary>
+        /// Tests deleting the author to book
+        /// Correct input
+        /// Returned identifier of adding pair
+        /// </summary>
+        [TestMethod]
+        public void RemoveAuthorFromBook_CorrectInput_ReturnedIdOfRemovingPair()
+        {
+            // Arrange
+            int bookId = 1;
+            int authorId = 1;
+            Book book = new Book { Id = 1, Name = "Test Book", Year = 1256 };
+            Author author = new Author { Id = 1, FullName = "Test author" };
+            List<BookAuthorPair> pairBookAuthor = new List<BookAuthorPair>()
```

Comment:
I noticed. that you often repeat this step, I mean creation of list for mock. You could add private method which create, fill in and returns this list. And attach it to mock then.
'private List<BookAuthorPair> GetPairBookAuthor()
{
declare list; 
initialize list;
return list
}; '
'mockDataProvider.SetupGet(mock => mock.PairsBookAuthor).Returns( GetPairBookAuthor());'

Reply:
Yes, I thought about it when I have already written all my tests...